web/api/v1/api.go

```diff
@@ -1813,20 +1816,208 @@ OUTER:
 	return matcherSets, nil
 }
 
+// marshalSeriesJSON writes something like the following:
+//
+//         {
+//            "metric" : {
+//               "__name__" : "up",
+//               "job" : "prometheus",
+//               "instance" : "localhost:9090"
+//            },
+//            "values": [
+//               [ 1435781451.781, "1" ],
+//               < more values>
+//            ],
+//            "histograms": [
+//               [ 1435781451.781, { < histogram, see below > } ],
+//               < more histograms >
+//            ],
+//         },
+func marshalSeriesJSON(ptr unsafe.Pointer, stream *jsoniter.Stream) {
+	s := *((*promql.Series)(ptr))
+	stream.WriteObjectStart()
+	stream.WriteObjectField(`metric`)
+	m, err := s.Metric.MarshalJSON()
+	if err != nil {
+		stream.Error = err
+		return
+	}
+	stream.SetBuffer(append(stream.Buffer(), m...))
+
+	// We make two passes through the series here: In the first marshaling
+	// all value points, in the second marshaling all histogram
+	// points. That's probably cheaper than just one pass in which we copy
+	// out histogram Points into a newly allocated slice for separate
+	// marshaling. (Could be benchmarked, though.)
+	foundSomething := false
+	for _, p := range s.Points {
```

Comment:
we could track a `histogramExists` in this loop so that we dont have to loop over all samples twice for every query, and just for results that contain histograms.

Reply:
Yes, good idea, will add.